File: DailyAssignmentProblems/DigitalDict.java
import java.io.*;
import java.util.*;
import java.util.Map;
import java.util.Scanner;
import java.util.Set;


class TestClass {
    static class FastReader {
        BufferedReader br;
        StringTokenizer st;

        public FastReader() {
            br = new BufferedReader(new
                     InputStreamReader(System.in));
        }

        String next() {
            while (st == null || !st.hasMoreElements()) {
                try {
                    st = new StringTokenizer(br.readLine());
                }
                catch (IOException  e) {
                    e.printStackTrace();
                }
            }
            return st.nextToken();
        }

        int nextInt() {
            return Integer.parseInt(next());
        }

        long nextLong() {
            return Long.parseLong(next());
        }

        double nextDouble() {
            return Double.parseDouble(next());
        }

        String nextLine() {
            String str = "";
            try {
                str = br.readLine();
            }
            catch (IOException e) {
                e.printStackTrace();
            }
            return str;
        }
    }
    public static class Node {
        Character data;
        boolean isTerminal;
        Node[] children;

        Node(Character data, boolean isTerminal) {
            this.data = data;
            this.isTerminal = isTerminal;
            this.children = new Node[26];
            for(int i=0;i<26;i++) {
                this.children[i]=null;
            }
        }
    }
    public static TestClass p = new TestClass();
    public Node root;
    public static PrintWriter pw = new PrintWriter(System.out);
    TestClass() {
        this.root = new Node('\0', false);
    }

    public static void addword(Node root, String s) {
        if (s.length() == 0) {
            if(!root.isTerminal) {
                root.isTerminal = true;
            }
            return;
        }
        Node child;
        int idx = s.charAt(0) - 'a';
        if (root.children[idx] != null) {
            child = root.children[idx];
        } else {
            child = new Node(s.charAt(0), false);
            root.children[idx] = child;
        }

        addword(child, s.substring(1));

    }

    public static void searchWord(Node parent, String word, String original) {
        if (word.length() == 0) {
            if(parent.isTerminal) {
                pw.println(original);
                pw.flush();
            }
            numberOfString(parent, original);
            return;
        }
        Character ch = word.charAt(0);
        int idx = word.charAt(0) - 'a';
        String ros = word.substring(1);
        Node child = parent.children[idx];
        if (child == null) {
            System.out.println("No suggestions");
            addword(p.root, original);
            return;
        }
        searchWord(child, ros, original);
    }

    public static int numberOfString(Node parent, String str) {
        if (parent == null) {
            return 0;
        }
        int result = 0;
        for (int i=0;i<parent.children.length;i++) {
            if(parent.children[i]==null) {
                continue;
            }
            if (parent.children[i].isTerminal) {
                pw.println(str + parent.children[i].data);
                pw.flush();
                result++;
            }
            result += numberOfString(parent.children[i], str + parent.children[i].data);
        }
        return result;
    }

    public static void main(String[] args) {
        FastReader scn = new FastReader();
        int n = scn.nextInt();

        while (n-- > 0) {
            addword(p.root, scn.next());
        }
        int q = scn.nextInt();
        while (q-- > 0) {
            String s=scn.next();
            searchWord(p.root, s, s);
        }
    }
}

File: src/Problems on Strings/DigitalDict.cpp
#include <iostream>
#include <map>
#define hashmap map<char, node *>
#define ll long long int
using namespace std;

class node
{
public:
  char data;
  hashmap h;
  bool isterminal;
  node(char ch)
  {
    data = ch;
    isterminal = false;
  }
};

class trie
{
public:
  node *root;
  trie()
  {
    root = new node('\0');
  }
  void addword(string word)
  {
    node *temp = root;
    char ch;
    for (int i = 0; word[i] != '\0'; i++)
    {
      ch = word[i];
      if (temp->h.count(ch) == 0)
      {
        node *child = new node(ch);
        temp->h[ch] = child;
        temp = child;
      }
      else
      {
        temp = temp->h[ch];
      }
    }
    temp->isterminal = true;
  }

  void print(node *temp, string word)
  {
    if (temp->h.empty())
    {
      cout << word << endl;
      return;
    }
    if (temp->isterminal)
    {
      cout << word << endl;
    }
    node *t;
    string w;
    for (auto it = temp->h.begin(); it != temp->h.end(); it++)
    {
      t = temp->h[it->first];
      w = word + it->first;
      print(t, w);
    }
    return;
  }

  bool search(string word)
  {
    node *temp = root;
    char ch;
    for (int i = 0; word[i] != '\0'; i++)
    {
      ch = word[i];
      if (temp->h.count(ch))
      {
        temp = temp->h[ch];
      }
      else
      {
        return false;
      }
    }
    print(temp, word);
    return true;
  }
};

int main()
{
  ll n, i;
  cin >> n;
  string s;
  trie t;
  for (i = 0; i < n; i++)
  {
    cin >> s;
    t.addword(s);
  }
  ll q, val;
  cin >> q;
  for (i = 0; i < q; i++)
  {
    cin >> s;
    val = t.search(s);
    /*if(val==false){
     // cout<<s<<"  "<<"No suggestions"<<endl;
    }
    else{
       // cout<<s<<" "<<"found"<<endl;
    }*/
  }
}
